cmd/govim/watcher.go

```diff
@@ -165,6 +87,59 @@ func (m *modWatcher) watch() {
 	}
 }
 
+// eventFilter return a function that is used to filter events according to the Go specific
+// rules described by https://golang.org/cmd/go/#hdr-Package_lists_and_patterns:
+//
+//  > Directory and file names that begin with "." or "_" are ignored by the go
+//    tool, as are directories named "testdata".
+//
+// and the module boundary described by https://golang.org/ref/mod#modules-overview:
+//
+//  > The module root directory is the directory that contains the go.mod file.
+//
+func eventFilter(root string) func(string) bool {
+	return func(path string) bool {
+		path = filepath.Clean(path)
+		if fi, err := os.Stat(path); err == nil && !fi.IsDir() {
+			var filename string
+			path, filename = filepath.Split(path)
+			switch filename[0] {
+			case '.', '_':
+				return true
+			}
+		}
+
+		// TODO: cache ignored directories to reduce amount of os.Stat(...) calls?
+		rel := strings.TrimPrefix(path, root)
+		rel = strings.Trim(rel, string(os.PathSeparator))
+		if rel == "" {
+			return false
+		}
+		parts := strings.Split(rel, string(os.PathSeparator))
+		if len(parts) == 0 {
+			return false
+		}
+
+		var curr string
+		for i := 0; i < len(parts); i++ {
+			dir := parts[i]
+			switch dir[0] {
+			case '.', '_':
+				return true
+			}
+			if dir == "testdata" {
+				return true
+			}
+
+			curr = filepath.Join(curr, dir)
+			if _, err := os.Stat(filepath.Join(root, curr, "go.mod")); err == nil {
+				return true
+			}
+		}
+		return false
+	}
+}
+
 func ofInterest(path string) bool {
 	return filepath.Ext(path) == ".go" || filepath.Base(path) == "go.mod" || filepath.Base(path) == "go.sum"
```

Comment:
I'm not sure this is entirely true in the context of CGO. Indeed the client can't really know what is relevant or not. And with `//go:embed` we could need to care about any file, because removal of a file that is the source of a `//go:embed` will result in an error.

So this would point to us having to notify about _everything_. Perhaps we address that in a follow up PR, along with batching the notifications?

Reply:
Interesting, this is unchanged code but that's a good point I didn't think about. Let's do it with the batching.